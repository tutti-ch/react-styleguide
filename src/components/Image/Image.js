import React, { Component } from "react";
import PropTypes from "prop-types";
import classes from "./Image.scss";
import classNames from "classnames";
import Spinner from "../Spinner";
import { TuttiCube } from "../../styles/Icons/assets/generic";
import { filterProps } from "../../helpers/functions";

// This gets overwritten by the Readme.md for some reason
// Apparently, it uses the component name as a global variable.
// Small hack to prevent this behavior.
const NativeImage =
  typeof window !== "undefined"
    ? window.Image
    : /* istanbul ignore next */ () => {};

/**
 * The Image component.
 *
 * Use this class instead of classic &lt;img&gt; tag.
 * It has built-in error handling and spinner in case the image takes long to load.
 */
export default class Image extends Component {
  static propTypes = {
    /**
     * The image source.
     */
    src: PropTypes.string.isRequired,

    /**
     * The size of the spinner while image is loading. By default this is disabled, set any number to activate.
     */
    spinnerSize: PropTypes.number,

    /**
     * The class name for the span that is shown on error.
     */
    errorClassName: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.array,
      PropTypes.object
    ]),

    /**
     * The class name for the spinner.
     */
    spinnerClassName: PropTypes.oneOfType([
      PropTypes.string,
      PropTypes.array,
      PropTypes.object
    ])
  };

  state = {
    loaded: false,
    error: false
  };

  /**
   * When the image is loaded set the state accordingly. This callback is triggered
   * by the Image class when its src property is updated.
   */
  onLoad = () => {
    this.setState({ error: false, loaded: true });
  };

  /**
   * When the image failed to load set the state accordingly. This callback is triggered
   * by the Image class when its src property is updated.
   */
  onError = () => {
    this.setState({ error: true, loaded: true });
  };

  componentDidMount() {
    // Why loading like this? Because we want to render only one element:
    // for error: true => we render the cube
    // for loaded: true => we render the image
    // otherwise the spinner.
    // Doing this without wrapping the elements in a span wrapper is almost impossible
    // and we do not want to wrap every image in the app with a span (breaks the layout in some cases)
    this.image = new NativeImage();
    this.image.onload = this.onLoad;
    this.image.onerror = this.onError;

    // Make sure to have this otherwise neither onload nor onerror will be triggered
    // and the image will spin forever.
    this.image.src = this.props.src;
  }

  // need to check both on update and on mount because sometimes the image gets called with an empty url at first
  // (eg. before we fetch the ad url from an api call)
  componentDidUpdate(oldProps) {
    if (oldProps.src !== this.props.src) {
      this.image.src = this.props.src;
      this.setState({ loaded: false });
    }
  }

  // Cancel callback functions in case we unmount
  // before the image has loaded. This will remove
  // the annoying setState warning.
  componentWillUnmount() {
    this.image.onload = undefined;
    this.image.onerror = undefined;
  }

  /**
   * Render the error state. This is called when onError callback is triggered.
   *
   * @return {*}
   */
  renderError = () => {
    const { errorClassName } = this.props;
    return (
      <span className={classNames(classes.error, errorClassName)}>
        <TuttiCube className={classes.cube} />
      </span>
    );
  };

  /**
   * Render the spinner. This is called when the image is still loading and
   * neither error, nor success is called yet.
   *
   * @return {*}
   */
  renderSpinner = () => {
    const { spinnerClassName, spinnerSize } = this.props;

    // Do not render if the spinner size is not provided
    if (!spinnerSize) {
      return null;
    }

    return <Spinner className={spinnerClassName} size={spinnerSize} />;
  };

  /**
   * Render the image component.
   *
   * @return {*}
   */
  render() {
    const { src } = this.props;
    const { error, loaded } = this.state;

    if (error) {
      return this.renderError();
    }

    if (loaded) {
      const attrs = filterProps(Image.propTypes, this.props);
      attrs.draggable =
        typeof attrs.draggable === "undefined" ? true : attrs.draggable; // By default is true
      attrs.className = classNames(classes.image, attrs.className);

      return (
        <img {...attrs} src={src} onLoad={this.onLoad} onError={this.onError} />
      );
    }

    return this.renderSpinner();
  }
}
