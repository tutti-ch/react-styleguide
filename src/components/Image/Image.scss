@import "../styles/main";

.image-wrapper {
  display: inline-block;
}

.image {
  display: inline-block;
}

.error {
  font-size: $size-xl;
  color: $greyscale-60;
}
