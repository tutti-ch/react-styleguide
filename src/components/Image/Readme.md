### Image component

```jsx
<Image src='http://thecatapi.com/api/images/get?format=src&type=gif' />
```
