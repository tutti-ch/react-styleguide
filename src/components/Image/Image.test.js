/* global describe, test, beforeEach, expect, jest */
import React from "react";
import Image from "components/Image";
import { mount } from "enzyme";
import toJson from "enzyme-to-json";

describe("(Component) Image", function() {
  let comp;
  const initialSrc = "https://tutti.ch/my-icon.jpg";

  beforeEach(function() {
    comp = mount(<Image src={initialSrc} />);
    comp.setState({ loaded: true });
  });

  test("snapshot should pass all props down", () => {
    const tree = mount(<Image src={initialSrc} />);
    const inst = tree.instance();
    inst.onLoad();
    expect(toJson(tree)).toMatchSnapshot();
    inst.onError();
    expect(toJson(tree)).toMatchSnapshot();
  });

  test("should render the image when loaded", () => {
    comp.setProps({ draggable: false });
    expect(comp.find("img[draggable]").exists()).toBe(true);
  });

  test("should render tutti cube on error", () => {
    // Simulate on error as the image-mock we have blocks
    // dom onError to be triggered.
    comp.instance().onError();
    comp.update();

    expect(comp.find("img").exists()).toBe(false);
    expect(comp.find("span.ico.ico-tutti-cube").exists()).toBe(true);

    // It should revert on load
    comp.instance().onLoad();
    comp.update();

    expect(comp.find("img").exists()).toBe(true);
    expect(comp.find("span.ico.ico-tutti-cube").exists()).toBe(false);
  });

  test("should always trigger callbacks", () => {
    const img = mount(<Image src="my-image" />);
    const inst = img.instance();
    expect(inst.state.loaded).toBe(true);
  });

  test("should trigger callbacks when the prop src is updated", () => {
    const instance = comp.instance();
    expect(instance.image.src).toBe(initialSrc);
    instance.componentWillUpdate({ src: "my-src.jpg" });
    expect(instance.image.src).toBe("my-src.jpg");

    // The result shoult be the same because we did not change the
    // props.src (it is still the initialOne)
    instance.componentWillUpdate({ src: initialSrc });
    expect(instance.image.src).toBe("my-src.jpg");
  });

  test("remove callbacks when component is unmounted", () => {
    const instance = comp.instance();
    expect(instance.image.onload).toBeInstanceOf(Function);
    expect(instance.image.onerror).toBeInstanceOf(Function);

    comp.unmount();
    expect(instance.image.onload).toBeUndefined();
    expect(instance.image.onerror).toBeUndefined();
  });

  test("shows a spinner if spinner size is defined", () => {
    comp.setState({ loaded: false });
    comp.setProps({ spinnerSize: 100 });
    comp.update();
    expect(comp.find("Spinner").length).toBe(1);
  });

  test("should render null when spinner is empty", () => {
    comp.setState({ error: false, loaded: false });
    expect(comp.instance().renderSpinner()).toBe(null);
  });
});
