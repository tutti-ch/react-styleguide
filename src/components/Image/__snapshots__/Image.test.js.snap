// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) Image snapshot should pass all props down 1`] = `
<Image
  src="https://tutti.ch/my-icon.jpg"
>
  <img
    className=""
    draggable={true}
    onError={[Function]}
    onLoad={[Function]}
    src="https://tutti.ch/my-icon.jpg"
  />
</Image>
`;

exports[`(Component) Image snapshot should pass all props down 2`] = `
<Image
  src="https://tutti.ch/my-icon.jpg"
>
  <img
    className=""
    draggable={true}
    onError={[Function]}
    onLoad={[Function]}
    src="https://tutti.ch/my-icon.jpg"
  />
</Image>
`;
